### PR TITLE
HPCC-21498 Write a tank file in ESP logging manager

### PR DESCRIPTION
**Pull request by wangkx**
Also fix bugs in filterLogContent():
1. Both logBackEndReq flag and logBackEndResp flag are filters;
2. The ESPLCGBackEndReq should not be in a filter loop because
it has been handled separately;
3. Renamed the hasFilters flag to foundGroupFilters;
4. The ESPContext tree itself already has the /ESPContext node
as the top tree node. If no filter for ESPLCGESPContext, the
existing code creates another /ESPContext node on the top of
the ESPContext tree before adding the tree to the logContentTree.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/logging/logginglib/compressutil.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "compressutil.hpp"
+
+#include "jlzw.hpp"
+
```
Comment (afishbeck) [inline esp/logging/logginglib/compressutil.cpp]: Note code copied from thcompressutil.cpp

Comment (afishbeck) [reply esp/logging/logginglib/compressutil.cpp]: Adds StringBuffer versions of routines.

File: esp/logging/loggingmanager/loggingmanager.cpp [hunk 2]
```diff
@@ -242,6 +259,70 @@ bool CLoggingManager::updateLog(IEspContext* espContext, IEspUpdateLogRequestWra
...
+    JBASE64_Encode(memBuf.toByteArray(), memBuf.length(), logData);
+
+    logData.append("</").append(LOGREQUEST).append(">");
+
+    return logData.length();
+}
+
+void CLoggingManager::addXMLItemToBuf(const char* tag, int value, StringBuffer& buf)
+{
+    buf.append("<").append(tag).append(">");
```
Comment (afishbeck) [inline esp/logging/loggingmanager/loggingmanager.cpp]: can use appendXMLOpenTag, appendXMLCloseTag functions.

File: esp/logging/loggingmanager/loggingmanager.cpp [hunk 3]
```diff
@@ -242,6 +259,70 @@ bool CLoggingManager::updateLog(IEspContext* espContext, IEspUpdateLogRequestWra
...
+    buf.append(value);
+    buf.append("</").append(tag).append(">");
+}
+
+void CLoggingManager::addXMLItemToBuf(const char* tag, const char* value, StringBuffer& buf)
```
Comment (afishbeck) [inline esp/logging/loggingmanager/loggingmanager.cpp]: Do you know that value does not need to be xml encoded?

File: esp/logging/loggingmanager/loggingmanager.cpp [hunk 4]
```diff
@@ -242,6 +259,70 @@ bool CLoggingManager::updateLog(IEspContext* espContext, IEspUpdateLogRequestWra
...
+{
```
Comment (afishbeck) [inline esp/logging/loggingmanager/loggingmanager.cpp]: could use appendXMLTag()

**Review by afishbeck — COMMENTED** [review]
@wangkx a few comments.

File: esp/logging/loggingmanager/loggingmanager.cpp (hunk 3 above)
Comment (wangkx) [reply esp/logging/loggingmanager/loggingmanager.cpp]: No need for the calls in this PR. But, I will add an option just in case of future use.

**Comment by wangkx** [comment]
@afishbeck please review my changes.

File: esp/logging/loggingmanager/loggingmanager.cpp [hunk 5]
```diff
@@ -291,36 +291,39 @@ bool CLoggingManager::saveToTankFile(IEspUpdateLogRequestWrap& logRequest, CLogR
 
 unsigned CLoggingManager::serializeLogRequestContent(IEspUpdateLogRequestWrap* request, const char* GUID, StringBuffer& logData)
 {
-    addXMLItemToBuf(LOGREQUEST_GUID, GUID, logData);
+    addXMLItemToBuf(LOGREQUEST_GUID, GUID, false, logData);
```
Comment (afishbeck) [inline esp/logging/loggingmanager/loggingmanager.cpp]: Haven't looked, so probably right, but why does this not need to be encoded?

**Review by afishbeck — COMMENTED** [review]
@wangkx just a question.

File: esp/logging/loggingmanager/loggingmanager.cpp (hunk 5 above)
Comment (wangkx) [reply esp/logging/loggingmanager/loggingmanager.cpp]: @afishbeck GUID is a random number + data/time. Does it need to be encoded?

**Comment by wangkx** [comment]
@afishbeck please review.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**